File: investigate/wikidata_preg.py
import requests
import random

def ciudad_random():

     # Consulta SPARQL para obtener ciudades de España
     # SPARQL es un lenguaje de consulta para conjuntos de datos

    consulta_sparql = """
    SELECT ?ciudad ?ciudadLabel ?codigoQ
    WHERE {
      ?ciudad wdt:P31 wd:Q515;
              wdt:P17 wd:Q29;
              wdt:P646 ?codigoQ.
      SERVICE wikibase:label { bd:serviceParam wikibase:language "[AUTO_LANGUAGE],es". }
    }
    """

    # URL de la API de Wikidata
    url_api_wikidata = "https://query.wikidata.org/sparql"

    # Encabezados para la solicitud HTTP
    headers = {
        "User-Agent": "EjemploCiudades/1.0",
        "Accept": "application/json"
    }

    # Realizar la solicitud HTTP con la consulta SPARQL
    respuesta = requests.get(url_api_wikidata, headers=headers, params={"query": consulta_sparql, "format": "json"})

    # Parsear la respuesta JSON
    datos = respuesta.json()

    # Extraer información
    ciudades = datos.get("results", {}).get("bindings", [])

    if ciudades:
        ciudad_aleatoria = random.choice(ciudades)
        nombre_ciudad = ciudad_aleatoria["ciudadLabel"]["value"]
        codigoQ = ciudad_aleatoria["codigoQ"]["value"]
        print(f"Ciudad aleatoria de España: {nombre_ciudad}")
    else:
        print("No se encontraron ciudades.")

        print(codigoQ + "codigoq")
        return codigoQ

def obtener_informacion_ciudad(nombre_ciudad_q):
    # Consulta a la API de Wikidata para obtener información sobre la ciudad
    #url = f"https://www.wikidata.org/w/api.php?action=wbgetentities&format=json&titles={nombre_ciudad}&sites=enwiki"
    #response = requests.get(url)
    #data = response.json()

    # Extraer el ID de la entidad de la ciudad desde la respuesta
    #city_id = list(data["entities"].keys())[0]

    # Consulta a la API de Wikidata para obtener información detallada sobre la ciudad
    url = f"https://www.wikidata.org/w/api.php?action=wbgetclaims&format=json&entity={nombre_ciudad_q}&property=P1082"
    response = requests.get(url)
    data = response.json()

    # Extraer la población de la ciudad desde la respuesta
    population_claim = data["claims"]["P1082"][0]["mainsnak"]["datavalue"]["value"]["amount"]
    return int(population_claim)

# Ejemplo de uso
nombre_ciudad_q = "Q14317"  # Puedes reemplazar con el nombre de la ciudad que estás buscando
poblacion = obtener_informacion_ciudad(nombre_ciudad_q)

# Plantilla para pregunta y respuesta
pregunta = f"¿Cuál es la población de {nombre_ciudad_q.capitalize()}?"
respuesta = f"La población de {nombre_ciudad_q.capitalize()} es {poblacion:,} habitantes."

print(pregunta)
print(respuesta)

print(ciudad_random())

File: investigate/wikidata_preg.js
import fetch from 'node-fetch';

// Código de ejemplo para generar preguntas usando la plantilla: ¿Cuál es la población de x? (dónde x es una ciudad aleatoria).

class GeneradorDePreguntasDeCiudades {

    // Función para obtener una ciudad aleatoria
    ciudadRandom() {
        const consultaSparql = `
            SELECT ?ciudad ?ciudadLabel ?codigoQ
            WHERE {
                ?ciudad wdt:P31 wd:Q515;
                        wdt:P17 wd:Q29;
                        rdfs:label ?ciudadLabel.
                SERVICE wikibase:label { bd:serviceParam wikibase:language "[AUTO_LANGUAGE],es". }
            }
        `;
    
        const urlApiWikidata = 'https://query.wikidata.org/sparql';
        const headers = {
            'User-Agent': 'EjemploCiudades/1.0',
            'Accept': 'application/json',
        };
    
        return fetch(`${urlApiWikidata}?query=${encodeURIComponent(consultaSparql)}&format=json`, {
            headers: headers,
        })
        .then(response => response.json())
        .then(datos => {
            const ciudades = datos.results.bindings;
    
            if (ciudades.length > 0) {
                const ciudadAleatoria = ciudades[Math.floor(Math.random() * ciudades.length)];
                const nombreCiudad = ciudadAleatoria.ciudadLabel.value;
                const codigoQ = ciudadAleatoria.ciudad.value.split('/').pop();
                return [nombreCiudad, codigoQ];
            } else {
                console.log('No se encontraron ciudades.');
                return null;
            }
        });
    }

    // Función para obtener la población de una ciudad
    obtenerPoblacionCiudad(codigoCiudad) {
        const url = `https://www.wikidata.org/w/api.php?action=wbgetclaims&format=json&entity=${codigoCiudad}&property=P1082`;
        return fetch(url)
        .then(response => response.json())
        .then(data => {
            const populationClaim = data.claims.P1082[0].mainsnak.datavalue.value.amount;
            return parseInt(populationClaim);
        });
    }

    // Función para obtener la población de una ciudad aleatoria con reintentos
    poblacionCiudadAleatoria() {
        const intentosMaximos = 25;
        let intentos = 0;
    
        const intentarObtenerPoblacion = () => {
            return this.ciudadRandom()
            .then(([nombreCiudad, codigoCiudad]) => {
                return this.obtenerPoblacionCiudad(codigoCiudad);
            })
            .catch(error => {
                intentos += 1;
                if (intentos < intentosMaximos) {
                    return intentarObtenerPoblacion();
                } else {
                    throw new Error('No se pudo obtener la población después de varios intentos.');
                }
            });
        };
    
        return intentarObtenerPoblacion();
    }

    // Función principal
    main() {
        return this.ciudadRandom()
        .then(([nombreCiudad, codigoQ]) => {
            return this.obtenerPoblacionCiudad(codigoQ)
            .then(poblacion => {
                const pregunta = `¿Cuál es la población de ${nombreCiudad.charAt(0).toUpperCase() + nombreCiudad.slice(1)}?`;
                const respuestaReal = `SOLUCION: La población de ${nombreCiudad.charAt(0).toUpperCase() + nombreCiudad.slice(1)} es ${poblacion.toLocaleString()} habitantes.`;
    
                return this.poblacionCiudadAleatoria()
                .then(respuestaA => {
                    const respuestaB = this.poblacionCiudadAleatoria();
                    const respuestaC = this.poblacionCiudadAleatoria();
                    const respuestaD = poblacion;
    
                    console.log(pregunta);
                    console.log(respuestaReal);
                    console.log("a)" + respuestaA);
                    console.log("b)" + respuestaB);
                    console.log("c)" + respuestaC);
                    console.log("d)" + respuestaD);
                });
            });
        });
    }

}

var generadorDePreguntas = new GeneradorDePreguntasDeCiudades();
generadorDePreguntas.main();